### PR TITLE
build: Use guix within gitian

### PR DESCRIPTION
**Pull request by achow101**
To somewhat ease the transition from gitian to guix, this PR replaces the previous gitian descriptors with ones that perform the guix build.

Due to the way guix works, some changes were needed to gitian itself, so builders must make sure that their gitian is up to date before trying this. Additionally, the base gitian images may need to be rebuilt.

I have only tested this using docker (with some changes to gitian: https://github.com/devrandom/gitian-builder/pull/251). It is my understanding that @luke-jr has been able to get these to work using kvm. I have tried with lxc but was unable to make it work.

The previous gitian descriptors used Ubuntu Focal as the image base. However because gitian is unable to make such images for lxc and kvm, the base image used here is Bionic.

### CODE REVIEW

**Comment by luke-jr** [comment]
I have failed to get a completed build so far, but have kept retrying.

It is very resource hungry, and the absence of apt-cacher means it has to restart downloads every time one fails.

I need to use my disksize patch for gitian to have enough space. https://github.com/devrandom/gitian-builder/pull/172

**Comment by luke-jr** [comment]
```
fab02ca332c5d79bcf1d62455a5e47e1b022522fb78ca81f2f07bde53d889a9f  aarch64-linux-gnu/bitcoin-21.99-guixtest1-aarch64-linux-gnu-debug.tar.gz
83c2e2a991bd2ae635e24667a1048e1735b77a887dc5a5f424f09b774c5bfb61  aarch64-linux-gnu/bitcoin-21.99-guixtest1-aarch64-linux-gnu.tar.gz
8d9c44cf254760000a8c62db00d919cc2e9488f6619b12123fb55a444870ce84  aarch64-linux-gnu/inputs.SHA256SUMS
db2d740831bc70cb443f2b855c076a905c99ed2567a3240d89c99f036dba1cf6  arm-linux-gnueabihf/bitcoin-21.99-guixtest1-arm-linux-gnueabihf-debug.tar.gz
92e3c65cacff6e522d8daf2bf71ca3eb2c62da67557990c4ee24ab9e17142a39  arm-linux-gnueabihf/bitcoin-21.99-guixtest1-arm-linux-gnueabihf.tar.gz
8d9c44cf254760000a8c62db00d919cc2e9488f6619b12123fb55a444870ce84  arm-linux-gnueabihf/inputs.SHA256SUMS
e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855  dist-archive/SKIPATTEST.TAG
10d67c880c5d1753ca2dc667241089d2e8a155282aabd17e483838239120cc87  dist-archive/bitcoin-21.99-guixtest1.tar.gz
cea5acf03c64d1f6958ee7ea5572b4ccbaee3525dde372f2e751e53bccd53137  powerpc64-linux-gnu/bitcoin-21.99-guixtest1-powerpc64-linux-gnu-debug.tar.gz
d71b0bc9f3ee82c5edf334547eb5c4383e51fc440192f95c8a1e0ab008668906  powerpc64-linux-gnu/bitcoin-21.99-guixtest1-powerpc64-linux-gnu.tar.gz
8d9c44cf254760000a8c62db00d919cc2e9488f6619b12123fb55a444870ce84  powerpc64-linux-gnu/inputs.SHA256SUMS
0e6cf8800b577a717177dedfc2f527522c8afbb1243688465747a0c1395883c9  powerpc64le-linux-gnu/bitcoin-21.99-guixtest1-powerpc64le-linux-gnu-debug.tar.gz
59d6d8f630d77676bcd88ecb28d8f6acd9e859ff16013b4b76e9033709755cd3  powerpc64le-linux-gnu/bitcoin-21.99-guixtest1-powerpc64le-linux-gnu.tar.gz
8d9c44cf254760000a8c62db00d919cc2e9488f6619b12123fb55a444870ce84  powerpc64le-linux-gnu/inputs.SHA256SUMS
e8e0622c8e748edaafeb38c7621980d9e29bebefe9fd25ff5dbd4d687e75ed08  riscv64-linux-gnu/bitcoin-21.99-guixtest1-riscv64-linux-gnu-debug.tar.gz
c45b797ff1092e015de43dfbeb8b911f0febc5d4c11f695d2ef9dc5118626059  riscv64-linux-gnu/bitcoin-21.99-guixtest1-riscv64-linux-gnu.tar.gz
8d9c44cf254760000a8c62db00d919cc2e9488f6619b12123fb55a444870ce84  riscv64-linux-gnu/inputs.SHA256SUMS
da94fccd6278cbd116d218f74150b0c1230bb40e1ac68bb0cba52503801b2366  x86_64-linux-gnu/bitcoin-21.99-guixtest1-x86_64-linux-gnu-debug.tar.gz
b925b9444131b1da245d5dd314040085f4dd9e9e513c28cfb6ae12fb0c5e6ba4  x86_64-linux-gnu/bitcoin-21.99-guixtest1-x86_64-linux-gnu.tar.gz
8d9c44cf254760000a8c62db00d919cc2e9488f6619b12123fb55a444870ce84  x86_64-linux-gnu/inputs.SHA256SUMS
5f79485f8299691166e860036149bc6d600b9d25db1e2d3a0a75a7bb18e4cbc3  x86_64-w64-mingw32/bitcoin-21.99-guixtest1-win-unsigned.tar.gz
157b6457bc8d4fafd680144fd684905a934985cdb8997e074c57b6134e5eb081  x86_64-w64-mingw32/bitcoin-21.99-guixtest1-win64-debug.zip
49ef90b5e9124203770c6c2dc154cbdab4098675fe4db93592bfef60eb31fb4a  x86_64-w64-mingw32/bitcoin-21.99-guixtest1-win64-setup-unsigned.exe
f4a8a7e079326a8500536eee5dfab38cb34a554f216a4aa2e7ab6d230b0dc813  x86_64-w64-mingw32/bitcoin-21.99-guixtest1-win64.zip
8d9c44cf254760000a8c62db00d919cc2e9488f6619b12123fb55a444870ce84  x86_64-w64-mingw32/inputs.SHA256SUMS
26f2f35486684b4b323359d11897a17e69a20a46a3524683f7e2bbc3391b1e07  bitcoin-core-guix-linux-win-res.yml
```

**Review by luke-jr — CHANGES_REQUESTED** [review]

File: contrib/gitian-descriptors/gitian-guix-linux-win.yml [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    sudo /var/guix/profiles/per-user/root/current-guix/bin/guix-daemon --build-users-group=guixbuild --substitute-urls="https://guix.carldong.io https://ci.guix.gnu.org"&
+  fi
+
+  # Import guix cache
+  if test -n "$GBUILD_CACHE_ENABLED"; then
+    mkdir -p ${GUIX_CACHE}
+    KEYFILE="${GUIX_CACHE}/signing-key.pub"
+    STORE_FILE="${GUIX_CACHE}/guix_store.nar"
+    if [[ -f "$KEYFILE" && -f "$STORE_FILE" ]]; then
+      sudo cat file | sudo guix archive --authorize
```
[the remark this answers is not in the document]
Comment (achow101) [reply contrib/gitian-descriptors/gitian-guix-linux-win.yml]: Oops.

`sudo` is not needed but there was a linter complaint about not having it.

File: contrib/gitian-descriptors/gitian-guix-linux-win.yml [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+files: []
+script: |
+  set -e -o pipefail
+
+  export TZ="UTC"
+  export BUILD_DIR="$PWD"
+  if test -n "$GBUILD_CACHE_ENABLED"; then
+    export SOURCES_PATH=${GBUILD_COMMON_CACHE}
+    export BASE_CACHE=${GBUILD_PACKAGE_CACHE}
+    mkdir -p ${BASE_CACHE} ${SOURCES_PATH}
```
[the remark this answers is not in the document]
Comment (achow101) [reply contrib/gitian-descriptors/gitian-guix-linux-win.yml]: Done

File: contrib/gitian-descriptors/gitian-guix-linux-win.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (luke-jr) [reply contrib/gitian-descriptors/gitian-guix-linux-win.yml]: Can we fix (or disable) the linter instead?

**Comment by luke-jr** [comment]
(Prefer to keep the old gitian yml removals in a separate PR)

**Comment by hebasto** [comment]
> (Prefer to keep the old gitian yml removals in a separate PR)

#22418

File: contrib/gitian-descriptors/gitian-guix-linux-win.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (achow101) [reply contrib/gitian-descriptors/gitian-guix-linux-win.yml]: We don't need sudo to read the keyfile, but we need to be able to pass it into `sudo guix archive`. The linter does not like `cat ${KEYFILE} | ...` and suggested that  a redirect be used (` ... < ${KEYFILE}`). But when I did that, it complained about the sudo and suggested this line.

**Comment by luke-jr** [comment]
@hebasto Yes, I mean *not* bundled in this PR

File: contrib/gitian-descriptors/gitian-guix-linux-win.yml [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+script: |
+  set -e -o pipefail
+
+  export TZ="UTC"
+  export BUILD_DIR="$PWD"
+  if test -n "$GBUILD_CACHE_ENABLED"; then
+    export SOURCES_PATH=${GBUILD_COMMON_CACHE}
+    export BASE_CACHE=${GBUILD_PACKAGE_CACHE}
+    mkdir -p "${BASE_CACHE}" "${SOURCES_PATH}"
+    GUIX_CACHE=${GBUILD_COMMON_CACHE}/guix
```
Comment (luke-jr) [inline contrib/gitian-descriptors/gitian-guix-linux-win.yml:29]:
```suggestion
    GUIX_CACHE="${GBUILD_COMMON_CACHE}/guix"
```

File: contrib/gitian-descriptors/gitian-guix-linux-win.yml [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+    GUIX_SCRIPT=${GUIX_CACHE}/guix-install.sh
```
Comment (luke-jr) [inline contrib/gitian-descriptors/gitian-guix-linux-win.yml:31]:
```suggestion
    GUIX_SCRIPT="${GUIX_CACHE}/guix-install.sh"
```

File: contrib/gitian-descriptors/gitian-guix-linux-win.yml [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    if [[ -f "${GUIX_SCRIPT}" ]]; then
+      cp ${GUIX_SCRIPT} $BUILD_DIR
```
Comment (luke-jr) [inline contrib/gitian-descriptors/gitian-guix-linux-win.yml:33]:
```suggestion
      cp "${GUIX_SCRIPT}" "$BUILD_DIR"
```

File: contrib/gitian-descriptors/gitian-guix-linux-win.yml [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    fi
+    GUIX_TARBALL=${GUIX_CACHE}/guix-tarball-pack-x86_64-linux.tar.gz
```
Comment (luke-jr) [inline contrib/gitian-descriptors/gitian-guix-linux-win.yml:35]:
```suggestion
    GUIX_TARBALL="${GUIX_CACHE}/guix-tarball-pack-x86_64-linux.tar.gz"
```

File: contrib/gitian-descriptors/gitian-guix-linux-win.yml [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    if [[ -f "${GUIX_TARBALL}" ]]; then
+      export GUIX_BINARY_FILE_NAME=${GUIX_TARBALL}
+    fi
+  fi
+  export JOBS=${NUM_PROCS}
+
+  # Install guix with the installer script
+  wget -nc https://git.savannah.gnu.org/cgit/guix.git/plain/etc/guix-install.sh
```
Comment (luke-jr) [inline contrib/gitian-descriptors/gitian-guix-linux-win.yml:43]: Can we verify the SHA256 of this? (and peg it to the current version)

File: contrib/gitian-descriptors/gitian-guix-linux-win.yml [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    # Not running, try starting it
+    # Not all virtualization methods have systemd available, so start guix-daemon manually
+    export GUIX_LOCPATH=/var/guix/profiles/per-user/root/guix-profile/lib/locale
+    export LC_ALL=en_US.utf8
+    sudo /var/guix/profiles/per-user/root/current-guix/bin/guix-daemon --build-users-group=guixbuild --substitute-urls="https://guix.carldong.io https://ci.guix.gnu.org"&
+  fi
+
+  # Import guix cache
+  if test -n "$GBUILD_CACHE_ENABLED"; then
+    mkdir -p ${GUIX_CACHE}
```
Comment (luke-jr) [inline contrib/gitian-descriptors/gitian-guix-linux-win.yml:61]:
```suggestion
    mkdir -p "${GUIX_CACHE}"
```

File: contrib/gitian-descriptors/gitian-guix-linux-win.yml [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    KEYFILE="${GUIX_CACHE}/signing-key.pub"
+    STORE_FILE="${GUIX_CACHE}/guix_store.nar"
+    if [[ -f "$KEYFILE" && -f "$STORE_FILE" ]]; then
+      sudo cat $KEYFILE | sudo guix archive --authorize
```
Comment (luke-jr) [inline contrib/gitian-descriptors/gitian-guix-linux-win.yml:65]:
```suggestion
      sudo cat "$KEYFILE" | sudo guix archive --authorize
```

File: contrib/gitian-descriptors/gitian-guix-linux-win.yml [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+      guix archive --import < ${STORE_FILE}
```
Comment (luke-jr) [inline contrib/gitian-descriptors/gitian-guix-linux-win.yml:66]:
```suggestion
      guix archive --import < "${STORE_FILE}"
```

File: contrib/gitian-descriptors/gitian-guix-linux-win.yml [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+  # Define DISTNAME variable.
+  # shellcheck source=contrib/gitian-descriptors/assign_DISTNAME
+  source contrib/gitian-descriptors/assign_DISTNAME
+
+  # Do the build
+  export HOSTS="x86_64-linux-gnu arm-linux-gnueabihf aarch64-linux-gnu riscv64-linux-gnu powerpc64-linux-gnu powerpc64le-linux-gnu x86_64-w64-mingw32"
+  ./contrib/guix/guix-build
+
+  # Move all of the guix output to our output dir
+  mv guix-build-${VERSION}/output/* ${OUTDIR}
```
Comment (luke-jr) [inline contrib/gitian-descriptors/gitian-guix-linux-win.yml:82]:
```suggestion
  mv "guix-build-${VERSION}/output"/* "${OUTDIR}"
```

File: contrib/gitian-descriptors/gitian-guix-linux-win.yml [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+
+  if test -n "$GBUILD_CACHE_ENABLED"; then
+    cd ${BUILD_DIR}
```
Comment (luke-jr) [inline contrib/gitian-descriptors/gitian-guix-linux-win.yml:85]:
```suggestion
    cd "${BUILD_DIR}"
```

File: contrib/gitian-descriptors/gitian-guix-linux-win.yml [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+
+    # Cache guix installer
+    cp guix-install.sh ${GUIX_SCRIPT}
```
Comment (luke-jr) [inline contrib/gitian-descriptors/gitian-guix-linux-win.yml:88]:
```suggestion
    cp guix-install.sh "${GUIX_SCRIPT}"
```

File: contrib/gitian-descriptors/gitian-guix-linux-win.yml [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+    guix pack --profile-name=current-guix --localstatedir guix
+    GUIX_PACK_TARBALL=$(find /gnu/store -name "*-tarball-pack.tar.gz" -type f)
+    sudo cp ${GUIX_PACK_TARBALL} ${GUIX_TARBALL}
```
Comment (luke-jr) [inline contrib/gitian-descriptors/gitian-guix-linux-win.yml:91]:
```suggestion
    sudo cp "${GUIX_PACK_TARBALL}" "${GUIX_TARBALL}"
```

File: contrib/gitian-descriptors/gitian-guix-linux-win.yml [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+
+    # Find all of the items in the guix store and export them to the cache
+    STORE_FILES=$(find /gnu/store -maxdepth 1 -mindepth 1 -type d -not -name ".links" | sort)
+    sudo guix archive --generate-key
+    sudo cp /etc/guix/signing-key.pub ${KEYFILE}
```
Comment (luke-jr) [inline contrib/gitian-descriptors/gitian-guix-linux-win.yml:96]:
```suggestion
    sudo cp /etc/guix/signing-key.pub "${KEYFILE}"
```

File: contrib/gitian-descriptors/gitian-guix-linux-win.yml [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+    guix archive --export --recursive $STORE_FILES > ${STORE_FILE}
```
Comment (luke-jr) [inline contrib/gitian-descriptors/gitian-guix-linux-win.yml:97]:
```suggestion
    guix archive --export --recursive "$STORE_FILES" > "${STORE_FILE}"
```

**Review by luke-jr — CHANGES_REQUESTED** [review]

File: contrib/gitian-descriptors/gitian-guix-linux-win.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (achow101) [reply contrib/gitian-descriptors/gitian-guix-linux-win.yml]: I disabled the linter warnings for these lines.

**Comment by achow101** [comment]
> (Prefer to keep the old gitian yml removals in a separate PR)

Dropped that commit

File: contrib/gitian-descriptors/gitian-guix-linux-win.yml (hunk 3 above)
Comment (achow101) [reply contrib/gitian-descriptors/gitian-guix-linux-win.yml]: Done here and elsewhere.

File: contrib/gitian-descriptors/gitian-guix-linux-win.yml (hunk 7 above)
Comment (achow101) [reply contrib/gitian-descriptors/gitian-guix-linux-win.yml]: Done

**Comment by fanquake** [comment]
Concept NACK - I think if someone would like to maintain these descriptors for some sort of transitional period, they are free to do so, but I'm not sure they need to live here. From what I can see we've already got a fairly solid number of builders for the 22.0 rcs we've had (especially given it's a new system), so Guix doesn't seem to be scaring anyone away. Will be interesting to see how many more we end up with for final.

Ideally we'll go ahead with removing gitian from this repo shortly: #22418.

**Comment by laanwj** [comment]
I agree with @fanquake above. It might make sense to maintain this somewhere else (could even be in the [docs](https://github.com/bitcoin-core/docs) repository which historically covers more specific building instructions), but I don't think it makes sense to have it (or anything referring to gitian) as part of this repository. It's not the recommended way to build.

**Comment by fanquake** [comment]
Going to close this for now then. #22418 has been merged as well. 

We currently have 13 builders who have submitted sigs for the `v22.0` tag, and I'm expecting a number more.

**Pull request closed without merging**